### PR TITLE
RFE-8788:Update installer to accept Capacity Group Reservation IDs for Azure

### PR DESCRIPTION
**Pull request by hlipsig**
To guarantee VM compute availability in Azure a Group Reservation ID can be used. This PR adds the ability to use this reservation at install time. 

## Summary by CodeRabbit

* **New Features**
  * Added support for configuring Azure Capacity Reservation Groups for the default platform, control plane, and compute pools.
  * Azure machine specifications now include the Capacity Reservation Group ID when configured.
  * Automatically defaults the group subscription ID from Azure credentials when omitted.

* **Validation**
  * Validates group existence, naming, required fields, region compatibility, and Azure Stack compatibility during provisioning.

* **Tests**
  * Added coverage for ID formatting, validation edge cases, missing groups, wrong regions, and naming rules.

### CODE REVIEW

**Comment by rna-afk** [comment]
We may need to add validation to check and see if the reservation actually exists in azure before we run the create cluster with wrong ID and have it report some error after like 20 mins of the installation. That would go into this file [here ](https://github.com/openshift/installer/blob/main/pkg/asset/installconfig/azure/validation.go) 

which would need a new function to get from Azure [here](https://github.com/openshift/installer/blob/main/pkg/asset/installconfig/azure/client.go)

which would need a mock function generated using 
```
mockgen -source=./client.go -destination=mock/azureclient_generated.go -package=mock
```

and added tests [here](https://github.com/openshift/installer/blob/main/pkg/asset/installconfig/azure/validation_test.go)

We also prefer to collect all the info and generate the id instead of passing the entire id by itself. Something like the customer managed key setup like [here](https://github.com/openshift/installer/blob/8beeb74ad9d29002e50f5530390c4caf5c2bf500/pkg/infrastructure/azure/storage.go#L122) and install config input like [this](https://github.com/openshift/installer/blob/8beeb74ad9d29002e50f5530390c4caf5c2bf500/pkg/types/azure/platform.go#L159)

**Comment by hlipsig** [comment]
> We may need to add validation to check and see if the reservation actually exists in azure before we run the create cluster with wrong ID and have it report some error after like 20 mins of the installation. That would go into this file [here ](https://github.com/openshift/installer/blob/main/pkg/asset/installconfig/azure/validation.go)
> 
> which would need a new function to get from Azure [here](https://github.com/openshift/installer/blob/main/pkg/asset/installconfig/azure/client.go)
> 
> which would need a mock function generated using
> 
> ```
> mockgen -source=./client.go -destination=mock/azureclient_generated.go -package=mock
> ```
> 
> and added tests [here](https://github.com/openshift/installer/blob/main/pkg/asset/installconfig/azure/validation_test.go)
> 
> We also prefer to collect all the info and generate the id instead of passing the entire id by itself. Something like the customer managed key setup like [here](https://github.com/openshift/installer/blob/8beeb74ad9d29002e50f5530390c4caf5c2bf500/pkg/infrastructure/azure/storage.go#L122) and install config input like [this](https://github.com/openshift/installer/blob/8beeb74ad9d29002e50f5530390c4caf5c2bf500/pkg/types/azure/platform.go#L159)

Thanks for the review! I'll make these added changes and then when I'm confident I'll make this not a WIP.

File: pkg/asset/machines/azure/machines.go [hunk 1]
```diff
@@ -253,6 +253,11 @@ func provider(platform *azure.Platform, mpool *azure.MachinePool, osImage string
 		dataDisks = append(dataDisks, dataDisk)
 	}
 
+	capacityReservationGroupID := ""
+	if mpool.CapacityReservationGroup != nil {
+		capacityReservationGroupID = mpool.CapacityReservationGroup.ToID()
+	}
```
[the remark this answers is not in the document]
Comment (hlipsig) [reply pkg/asset/machines/azure/machines.go:259]: false positive.  machineapi.AzureMachineProviderSpec.CapacityReservationGroupID  is a  string , not  *string . CodeRabbit confused it with the CAPZ type.

File: pkg/types/azure/validation/capacityreservation_test.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+		{
+			name: "invalid name",
+			group: func() *azure.CapacityReservationGroup {
+				group := valid()
+				group.Name = ""
+				return group
+			}(),
+			cloudName: azure.PublicCloud,
+			expected:  `capacityReservationGroup.name: Invalid value: "": invalid capacity reservation group name format`,
+		},
```
[the remark this answers is not in the document]
Comment (hlipsig) [reply pkg/types/azure/validation/capacityreservation_test.go:83]: Will do this.

**Comment by hlipsig** [comment]
Note for the reviewer. This work was done on a windows machine so there could be traces of line ending encoding specific to the windows environment. If that's so I will fix next week when I have my Linux machine back.

**Comment by jhixson74** [comment]
This LGTM. I also did a git diff locally and did not see any strange character return/ line feed issues. I'll wait for you to return before prematurely approving/lgtm'ing.

**Comment by hlipsig** [comment]
> This LGTM. I also did a git diff locally and did not see any strange character return/ line feed issues. I'll wait for you to return before prematurely approving/lgtm'ing.

Looks good to me as well. Thanks

**Comment by hlipsig** [comment]
Making sure this doesn't merge as non-critical for 5.0. Will also update to make sure CAPI works and submit evidence of manual test

**Comment by hlipsig** [comment]
/verified @hlipsig 


AZ CLI Out put from master nodes during install test:  
```
crgtest08101810-dr47b-master-0   1     crgtest08101810-crg                                                        
 crgtest08101810-dr47b-master-1   1     crgtest08101810-crg                                                        
 crgtest08101810-dr47b-master-2   1     crgtest08101810-crg      
```                                                
Bootstrap also referenced the same CRG:                                                                          
` crgtest08101810-dr47b-bootstrap  1     crgtest08101810-crg  `

**Comment by rna-afk** [comment]
Not related to this PR, just running it to see if it works

/test e2e-azurestack

**Comment by hlipsig** [comment]
Screenshot 2026-08-17 at 3 08 34 PM
Screenshot 2026-08-17 at 3 08 43 PM
Pictures of the  Capacity Reservation on 1 worker and 1 Controlplane node in a 6 node cluster install.
